### PR TITLE
Chaishi/websocket

### PR DESCRIPTION
**Pull request by changchaishi**

### CODE REVIEW

File: pt/app.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 from config import db
 from endpoints import RESOURCES
 
+from utils.socketio import socketio
```
Comment (changchaishi) [inline pt/app.py:11]: socket is created at pt/utils/socketio.py file

File: pt/endpoints/socketio/resource.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+class SocketResource(Resource):
+    # pylint: disable=R0201
+    def post(self):
+        socketio.emit('transaction', get_settlement())
+        logging.info(
+            "[Emit Settlement Transactions]\nMessage sent"
+        )
+        return jsonify({"msg": "Successfulling emitted settlement transactions"})
+    # pylint: enable=R0201
```
Comment (changchaishi) [inline pt/endpoints/socketio/resource.py:17]: In order to ask socket server ( the flask server ) to emit message. An API route should be added for calling

File: pt/scripts/settlement.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        }
+    ]
+    now = datetime.now()
+    data = {
+        "date": now.strftime("%Y/%m/%d"),
+        "time": now.strftime("%H:%M"),
+        "transactions": transactions
+    }
+
+    return json.dumps(data)
```
Comment (changchaishi) [inline pt/scripts/settlement.py]: This function is solely for generating test settlement for front-end use case. If we start working on true settlement calculation, we can do it in this function too, in the future

File: pt/utils/socketio.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            logging.info(
+                "[SocketIO]\nShort-lived token expired or invalid"
+            )
+    else:
+        logging.info(
+            "[SocketIO]\nShort-lived token not provided"
+        )
+
+    if verified is False:
+        raise ConnectionRefusedError("Unauthorized access")
```
Comment (changchaishi) [inline pt/utils/socketio.py:56]: This file is about creating socketio and listening events. Authentication is also taking place in this file.

File: pt/app.py (hunk 1 above)
Comment (Sirius207) [reply pt/app.py:11]: You can write down these comments in PR description, not in a review conversation.

File: pt/endpoints/socketio/resource.py (hunk 2 above)
Comment (Sirius207) [reply pt/endpoints/socketio/resource.py:17]: This comment should be written inside this file.

File: pt/endpoints/socketio/resource.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# from datetime import date
```
Comment (Sirius207) [inline pt/endpoints/socketio/resource.py]: Maybe you can remove this line if it is not necessary.

File: pt/scripts/settlement.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+from datetime import datetime
+
+
+def get_settlement():
+    achievement_list = list(range(0, 105, 5))
+    transactions = [
+        {
+            "seller": "Carlab_BEMS",
+            "buyer": "SGESC_C_BEMS",
+            "achievement": random.choice(achievement_list) / 100
```
Comment (Sirius207) [inline pt/scripts/settlement.py]: What is the meaning of Carlab_BEM and SSGESC_C_BEMS?

File: pt/scripts/settlement.py (hunk 3 above)
Comment (Sirius207) [reply pt/scripts/settlement.py]: OK, maybe you can write this purpose on top of this file

How about renaming this file to test_settlement.py?

File: pt/utils/socketio.py (hunk 4 above)
Comment (Sirius207) [reply pt/utils/socketio.py:56]: Maybe you can write this purpose on top of this file.

File: pt/utils/socketio.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+
+# pylint: disable=C0103
+socketio = SocketIO()
+# pylint: enable=C0103
+
+
+@socketio.on('connect')
+def connected():
+    verified = False
```
Comment (Sirius207) [inline pt/utils/socketio.py:29]: Maybe is_verified is more accurate to its meaning

Like [this](https://stackoverflow.com/questions/18500718/fql-user-table-is-verified-vs-verified)

**Comment by changchaishi** [comment]
This pull request consists of implementations of socketio and socket test.

For socketio to work on the server. I have to put socketio object to another file named 'socketio.py', it instantiate the socket object and define socket events. So after instantiation, socketio can bind in 'app.py'.

For socket testing ( Dump sample data to front-end client ), back-end server needs to emit 'test_settlement' periodically. By doing this emit action, another API route called (/socket_settlement) should be added to emit sample data.

File: pt/app.py (hunk 1 above)
Comment (changchaishi) [reply pt/app.py:11]: wrote down. Fixed

File: pt/scripts/settlement.py (hunk 3 above)
Comment (changchaishi) [reply pt/scripts/settlement.py]: fixed, changed the file name to test_settlement.py

File: pt/utils/socketio.py (hunk 4 above)
Comment (changchaishi) [reply pt/utils/socketio.py:56]: Fixed it, I wrote down doc-type string as block comment for this file.

File: pt/endpoints/socketio/resource.py (hunk 5 above)
Comment (changchaishi) [reply pt/endpoints/socketio/resource.py]: Line removed.

File: pt/scripts/settlement.py (hunk 6 above)
Comment (changchaishi) [reply pt/scripts/settlement.py]: Fixed, added a doc-type string on the top of function.

File: pt/utils/socketio.py (hunk 7 above)
Comment (changchaishi) [reply pt/utils/socketio.py:29]: Fixed. renamed the function to is_verified()

**Comment by changchaishi** [comment]
All issues and bugs fixed.

**Review by Sirius207 — APPROVED** [review]
OK, looks good.

**Pull request merged**